back-end/routes/clusters.js

```diff
@@ -0,0 +1,208 @@
+const express = require('express');
+const { PrismaClient } = require('@prisma/client');
+const fetch = require('node-fetch');
+const router = express.Router();
+const prisma = new PrismaClient();
+
+router.use(express.json());
+
+const fetchAudioFeatures = async (songs, accessToken) => {
+    const songIds = songs.map(song => (song.uri).split(':').pop());
+    const stringOfIds = songIds.join(',');
+    try {
+        const response = await fetch(`https://api.spotify.com/v1/audio-features?ids=${encodeURIComponent(stringOfIds)}`, {
+            headers: {
+                'Authorization': `Bearer ${accessToken}`
+            }
+        });
+        if (!response.ok) {
+            throw new Error(`HTTP error! status: ${response.status}`);
+        }
+        const audioData = await response.json();
+        return audioData.audio_features;
+    } catch (error) {
+        console.error(error);
+        return [];
+    }
+};
+
+const getAccessToken = async (userId) => {
+    const token = await prisma.token.findFirst({
+        where: {
+            userId
+        },
+        orderBy: {
+            createdAt: 'desc'
+        }
+    });
+    return token ? token.token : null;
+};
+
+function getValenceFeatures(audioFeatures) {
+    const valenceFeatures = [];
+    audioFeatures.forEach(feature => {
+        const songId = feature.uri.split(':').pop();
+        const valenceValue = feature.valence;
+        valenceFeatures.push({ id: songId, valence: valenceValue });
+    });
+    return valenceFeatures;
+}
+
+async function fetchAndProcessAudioFeatures(songs, accessToken) {
+    const features = await fetchAudioFeatures(songs, accessToken);
+    const valenceFeatures = getValenceFeatures(features);
+    return valenceFeatures;
+}
+
+function euclideanDistance(a, b) {
+    return Math.abs(a - b);
+}
+
+// Calculates mean of valence value of a cluster
+function calculateMean(cluster) {
```

Comment:
Can we probably rename to `calculateMeanValence`?

Reply:
For reusability for later stretch goal of adding different clustering value, name can stay the same, just different way of use that takes whatever "data value" instead of "valence" and finds mean of cluster.